oslabs-beta/watchmo
# watchmo watch command

The watchmo watch command is now functional!

**HOW IT WORKS**
Watchmo watch uses process.hrtime, Promises, and async/await to query the GraphQL server and time the responses. 

watchmo init sets up a config.json file in the working directory of a new project. The data in this config file is passed as arguments to watchmo watch through yargs in watchmo.js. The current structure of the config is as follows: 
```
{
  "endpoint": '',//The graphQL endpoint
  "categories: {
      //category names
      "default" : {
          "queries": [], //The graphQL queries to be watched
          "frequency": -1, //The wait time in ms between queries
       }
   }
}
```

watchmo watch establishes a setInterval for each category using the frequency information associated with that category, and calls an asynchronous function "sendQueriesAndSave".

"sendQueriesAndSave" function loops through each query and awaits an associated query Promise created by "buildQueryPromise", then saves the resolved data as a new file to a pathname given as a parameter to watch in watchmo.js.

"buildQueryPromise" returns a Promise that starts a timer, sends the given GQL query, then ends the timer and resolves to an object containing the response and the timing information. Note that since "buildQueryPromise: is awaited in "sendQueriesAndSave", this timing information is not corrupted by any other threads of execution.

In summary, watchmo watch takes the data in the config file, uses setIntervals to call a function that awaits a Promise that sets a timer for each query, and finally saves the data in a new file to a given path. 

**HOW TO TEST IT**:
 _Instructions for setting up the command line for watchmo in general_ 
1. In the watchmo project directory, run `npm link` (This does not need to be re-run after it has been run once). Run `watchmo --help` to ensure watchmo CLI is working. It shouldn't return an error. 
2. make a new project directory
3. cd into the project directory
4. run `watchmo init` to setup the watchmo directory.

 _Instructions for setting up the GQL test server (courtesy of Evan Hilton)_ 
1. clone Evan's repo: [GraphQLDummyServer](https://github.com/EH1537/GraphQLDummyServer) 
2. cd into the cloned project directory
3. run `npm run start` to start the server. The console should say
- 'now listening on port 4545'
- 'Connected to Mongo DB for Books (AZURE IN VIRGINIA).'
- 'Connected to Mongo DB for Authors. (AWS IN Ireland)'
- Possibly other comments if the server has been updated.
4. In your browser navigate to [http://localhost:4545/graphql](http://localhost:4545/graphql). You should see an interface that lets you make queries to the GQL server directly. Feel free to play with this if you want to get familiar with the structure of GQL queries.

_Instructions for testing watchmo watch_ 
1. Make sure that you are in the root of a testing Project directory, NOT the watchmo project directory.
2. Check that you have a config.json file. If not, go to the root directory of your testing project and run `watchmo init`.
3. Edit the config.json file to have the appropriate information. Mine looks like this:
```
{
  "endpoint": "http://localhost:4545/graphql",
  "categories": {
    "default": {
      "queries": [
        "{ books { name }}",
        "{ book(id: \"5e28f0b98ff7a733f8dae1ad\") { name }}",
        "{ authors { name }}"
      ],
      "frequency": 5000
    },
    "test": {
      "queries": [
        "{ authors { name }}"
      ],
      "frequency": 5500
    }
  }
}
```
4. In the testing project root directory, run `watchmo watch`. After an amount of time specified by the "frequency" values in the config file, the console should log 'saving file to (file path)', and in your testing project's "/watchmo/snapshots" directory you should see new files being created at regular intervals.
